pymc3/gp/gp.py

```diff
@@ -1,151 +1,308 @@
 import numpy as np
-from scipy import stats
-from tqdm import tqdm
 
-from theano.tensor.nlinalg import matrix_inverse
+import theano
 import theano.tensor as tt
+import theano.tensor.slinalg
 
-from .mean import Zero, Mean
-from .cov import Covariance
-from ..distributions import MvNormal, Continuous, draw_values, generate_samples
-from ..model import modelcontext
+import pymc3 as pm
 
+__all__ = ['GPLatent', 'GPMarginal', 'TProcess', 'GPMarginalSparse']
 
-__all__ = ['GP', 'sample_gp']
+cholesky = pm.distributions.dist_math.Cholesky(nofail=True, lower=True)
+solve_lower = tt.slinalg.Solve(A_structure='lower_triangular')
+solve_upper = tt.slinalg.Solve(A_structure='upper_triangular')
 
+def stabilize(K):
+    """ adds small diagonal to a covariance matrix """
+    return K + 1e-6 * tt.identity_like(K)
 
-class GP(Continuous):
-    """Gausian process
 
-    Parameters
-    ----------
-    mean_func : Mean
-        Mean function of Gaussian process
-    cov_func : Covariance
-        Covariance function of Gaussian process
-    X : array
-        Grid of points to evaluate Gaussian process over. Only required if the
-        GP is not an observed variable.
-    sigma : scalar or array
-        Observation standard deviation (defaults to zero)
+class GPBase(object):
     """
-    def __init__(self, mean_func=None, cov_func=None, X=None, sigma=0, *args, **kwargs):
+    Base class
+    """
+    def __init__(self, cov_func):
+        self.cov_func = cov_func
+        # force user to run `__call__` before `conditioned_on`
+        self._ready = False
+
+    def __add__(self, other):
+        if not isinstance(self, type(other)):
+            raise ValueError("cant add different GP types")
+        cov_func = self.cov_func + other.cov_func
+        return type(self)(cov_func)
+
+    def __call__(self, name, size, mean_func):
+        """Set state of GP constructor, separate function parameters from
+        conditioning variables (data, random variables)
+        This helps syntax, makes code clearer when defining likelihood, then
+        predictive distribution off same gp object.
+        """
+        self.name = name
+        self.size = size
+        self.mean_func = mean_func
 
-        if mean_func is None:
-            self.M = Zero()
+        # force user to run `__call__` before `conditioned_on`
+        self._ready = True
+        return self
+
+    def conditioned_on(self, X, Xs=None, **kwargs):
+        if self._ready:
+            self._ready = False
```

Comment:
Hmm that's true, the `_ready` check is brittle to that.  It's mostly there so that an informative error msg is raised when `__call__` isn't run before `conditioned_on`.  If it weren't there, I think the user would end up getting runtime errors when trying to access stuff like self.size, or when trying to construct a second random variable with the same name.